### PR TITLE
[TextControls] Add MDCFilledTextField theming extension

### PR DESCRIPTION
**Pull request by andrewoverton**
This PR adds the theming extension for MDCFilledTextField.

Here's a screenshot:

Screen Shot 2019-11-19 at 9 54 59 AM

Related to #8682.

### CODE REVIEW

File: MaterialComponents.podspec [hunk 1]
```diff
@@ -1906,6 +1906,25 @@ Pod::Spec.new do |mdc|
...
 
+  mdc.subspec "TextControls+Theming" do |extension|
+    extension.ios.deployment_target = '9.0'
+    extension.public_header_files = "components/#{extension.base_name.split('+')[0]}/src/#{extension.base_name.split('+')[1]}/*.h"
+    extension.source_files = [
+      "components/#{extension.base_name.split('+')[0]}/src/#{extension.base_name.split('+')[1]}/*.{h,m}",
+      "components/#{extension.base_name.split('+')[0]}/src/#{extension.base_name.split('+')[1]}/private/*.{h,m}"
+    ]
+    extension.dependency "MaterialComponents/#{extension.base_name.split('+')[0]}"
+    extension.dependency "MaterialComponents/Typography"
```
Comment (codeman7) [inline MaterialComponents.podspec]: Shouldn't this also include `schemes/Typography` and `schemes/Color`?

File: components/TextControls/BUILD [hunk 2]
```diff
@@ -47,24 +48,38 @@ mdc_public_objc_library(
...
         "//components/private/Math",
+        "@material_internationalization_ios//:MDFInternationalization",
+    ],
+)
+
+mdc_extension_objc_library(
+    name = "Theming",
+    deps = [
+        ":TextControls",
+        ":private",
```
Comment (codeman7) [inline components/TextControls/BUILD]: Why is `private` needed, shouldn't we only be theming public things? Also don't we need the subsystem schemes?

File: components/TextControls/src/Theming/MDCFilledTextField+MaterialTheming.h [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+ */
+- (void)applyThemeWithScheme:(nonnull id<MDCContainerScheming>)scheme;
+
+/**
+ Applies a container scheme's subsystem-specific schemes to the receiver in a manner best suited to
+ convey an error state.
+
+ @param scheme A container scheme instance.
+ */
+- (void)applyErrorThemeWithScheme:(nonnull id<MDCContainerScheming>)scheme;
```
Comment (codeman7) [inline components/TextControls/src/Theming/MDCFilledTextField+MaterialTheming.h:38]: If a user enters _invalid_ text then they (the client) should re-apply theme?

File: components/TextControls/src/Theming/MDCFilledTextField+MaterialTheming.m [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#import "MDCFilledTextField+MaterialTheming.h"
+
+#import <Foundation/Foundation.h>
+
+#import "../../src/private/MDCBaseTextField+MDCTextControl.h"
```
Comment (codeman7) [inline components/TextControls/src/Theming/MDCFilledTextField+MaterialTheming.m]: I don't think we should be theming `private` things.

File: components/TextControls/src/Theming/MDCFilledTextField+MaterialTheming.m [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+  [self applyDefaultColorScheme:[self colorSchemeWithContainerScheme:containerScheme]];
+}
+
+- (void)applyErrorThemeWithScheme:(nonnull id<MDCContainerScheming>)containerScheme {
+  [self applyTypographyScheme:[self typographySchemeWithContainerScheme:containerScheme]];
+  [self applyErrorColorScheme:[self colorSchemeWithContainerScheme:containerScheme]];
+}
+
+- (id<MDCColorScheming>)colorSchemeWithContainerScheme:
+    (nonnull id<MDCContainerScheming>)containerScheme {
```
Comment (codeman7) [inline components/TextControls/src/Theming/MDCFilledTextField+MaterialTheming.m]: Why is this needed? Isn't color and typography `nonnull`?

File: components/TextControls/src/Theming/MDCFilledTextField+MaterialTheming.m [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+
+- (void)applyDefaultColorScheme:(id<MDCColorScheming>)colorScheme {
+  CGFloat disabledOpacity = (CGFloat)0.60;
+
+  UIColor *textColorNormal = [colorScheme.onSurfaceColor colorWithAlphaComponent:(CGFloat)0.87];
+  UIColor *textColorEditing = textColorNormal;
+  UIColor *textColorDisabled = [textColorNormal colorWithAlphaComponent:disabledOpacity];
+
+  UIColor *assistiveLabelColorNormal =
+      [colorScheme.onSurfaceColor colorWithAlphaComponent:(CGFloat)0.60];
```
Comment (codeman7) [inline components/TextControls/src/Theming/MDCFilledTextField+MaterialTheming.m:45]: Nit: Define this alpha values in a const?

File: components/TextControls/src/Theming/MDCFilledTextField+MaterialTheming.h (hunk 3 above)
Comment (andrewoverton) [reply components/TextControls/src/Theming/MDCFilledTextField+MaterialTheming.h:38]: Yes. This was something @jverkoey @randallli and @romoore all felt very strongly about. The error "state" is not really a state, it's just a different application of a scheme. One benefit of this approach is it simplifies the implementation of overlapping "states," i.e. error coexisting with editing, resting, and disabled.

File: components/TextControls/src/Theming/MDCFilledTextField+MaterialTheming.m (hunk 5 above)
Comment (andrewoverton) [reply components/TextControls/src/Theming/MDCFilledTextField+MaterialTheming.m]: I think I wrote this before that change was made haha

File: components/TextControls/src/Theming/MDCFilledTextField+MaterialTheming.m (hunk 6 above)
Comment (andrewoverton) [reply components/TextControls/src/Theming/MDCFilledTextField+MaterialTheming.m:45]: Done.

File: components/TextControls/src/Theming/MDCFilledTextField+MaterialTheming.m (hunk 5 above)
Comment (andrewoverton) [reply components/TextControls/src/Theming/MDCFilledTextField+MaterialTheming.m]: Changed it.

File: components/TextControls/src/Theming/MDCFilledTextField+MaterialTheming.m (hunk 4 above)
Comment (andrewoverton) [reply components/TextControls/src/Theming/MDCFilledTextField+MaterialTheming.m]: Removed private imports.

File: components/TextControls/BUILD (hunk 2 above)
Comment (andrewoverton) [reply components/TextControls/BUILD]: Removed this dependency

File: MaterialComponents.podspec (hunk 1 above)
Comment (andrewoverton) [reply MaterialComponents.podspec]: Done.

File: components/TextControls/src/Theming/MDCFilledTextField+MaterialTheming.m (hunk 6 above)
Comment (codeman7) [reply components/TextControls/src/Theming/MDCFilledTextField+MaterialTheming.m:45]: Can we define all these values as constants?

File: components/TextControls/tests/unit/Theming/MDCFilledTextFieldThemingTests.m [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+@property(nonatomic, strong) MDCSemanticColorScheme *colorScheme;
+@property(nonatomic, strong) MDCTypographyScheme *typographyScheme;
+@property(nonatomic, strong) MDCContainerScheme *containerScheme;
+@end
+
+@implementation MDCFilledTextFieldThemingTest
+
+- (void)setUp {
+  [super setUp];
+  self.textField = [[MDCFilledTextField alloc] init];
```
Comment (codeman7) [inline components/TextControls/tests/unit/Theming/MDCFilledTextFieldThemingTests.m:32]: Nit: New line after `super`

File: components/TextControls/tests/unit/Theming/MDCFilledTextFieldThemingTests.m [hunk 8]
```diff
@@ -0,0 +1,324 @@
...
+  self.colorScheme =
+      [[MDCSemanticColorScheme alloc] initWithDefaults:MDCColorSchemeDefaultsMaterial201804];
+  self.typographyScheme =
+      [[MDCTypographyScheme alloc] initWithDefaults:MDCTypographySchemeDefaultsMaterial201804];
```
Comment (codeman7) [inline components/TextControls/tests/unit/Theming/MDCFilledTextFieldThemingTests.m:36]: Should we default these values to the latest enum values?

File: components/TextControls/tests/unit/Theming/MDCFilledTextFieldThemingTests.m [hunk 9]
```diff
@@ -0,0 +1,324 @@
...
+
+#import <XCTest/XCTest.h>
+
+#import "MaterialContainerScheme.h"
+#import "MaterialTextControls+Theming.h"
+#import "MaterialTextControls.h"
+
+@interface MDCFilledTextFieldThemingTest : XCTestCase
+@property(nonatomic, strong) MDCFilledTextField *textField;
+@property(nonatomic, strong) MDCSemanticColorScheme *colorScheme;
```
Comment (codeman7) [inline components/TextControls/tests/unit/Theming/MDCFilledTextFieldThemingTests.m:23]: Why do we need these variables?

**Review by codeman7 — APPROVED** [review]

File: components/TextControls/tests/unit/Theming/MDCFilledTextFieldThemingTests.m (hunk 7 above)
Comment (andrewoverton) [reply components/TextControls/tests/unit/Theming/MDCFilledTextFieldThemingTests.m:32]: Will add in a follow up.

File: components/TextControls/tests/unit/Theming/MDCFilledTextFieldThemingTests.m (hunk 8 above)
Comment (andrewoverton) [reply components/TextControls/tests/unit/Theming/MDCFilledTextFieldThemingTests.m:36]: Will change in a follow up.

File: components/TextControls/tests/unit/Theming/MDCFilledTextFieldThemingTests.m (hunk 9 above)
Comment (andrewoverton) [reply components/TextControls/tests/unit/Theming/MDCFilledTextFieldThemingTests.m:23]: I just prefer saying `self.colorScheme` over `self.containerScheme.colorScheme` etc.
